File: grafff.js
function Graph1() {
  const g = [];

  const edge = (a, b) => g.push([a, b]);
	const from = x => g.filter(([a, _]) => a === x).map(([_, b]) => b);
	const to = x => g.filter(([_, b]) => b === x).map(([a, _]) => a);

  return {
	  edge,
	  from,
	  to,
  };

}

function Graph3() {
  const g = new Map();

  const edge = (a, b) => {
    if (!g.has(a)) {
      g.set(a, []);
    }

    if (!g.get(a).find(x => x === b)) {
      g.get(a).push(b);
    }
  };

	const from = x => g.get(x) || [];

  const to = x => {
    const tos = [];

    for (let value of g) {
      const [a, b] = value;

      if (b.find(v => v === x)) {
        tos.push(a);
      }
    }

    return tos;
  };

  return {
	  edge,
	  from,
	  to,
    id: () => g
  };

}

var g2 = Graph3();
g2.edge(1,2);
g2.edge(2,4);
g2.edge(2,8);
g2.edge(8,9);
g2.edge(9,13);
g2.edge(13, 1);

let _travel = (start, beenThere = new Set()) => (vertex, g) => {
  if (g.has(vertex)) {
    g.get(vertex).forEach(
      e => {
        if (!beenThere.has(e) && e !== start) {
          beenThere.add(e);
          _travel(start, beenThere)(e, g);
        }
      }
    );
  }

  return beenThere;
};

let travel = (vertex, g) => _travel(vertex)(vertex, g);

const _findCycle3 = (g, a) => {
  if (!g.has(a)) {
    return false;
  };

  return true;
};

const edges3 = g => [...g.entries()].map(
  ([key, values]) => values.map(val => [key, val])
).reduce((a, b) => a.concat(b), []);

const isCyclic3 = g => {

};

function DAG3() {
}

File: grafff.clj
(ns grafff.core)

(defn start-with-edge [a b]
  {a (conj nil b)})

(defn add-edge [g a b]
  (update g a #(conj % b)))

(defn edge
  ([a b] (start-with-edge a b))
  ([g a b] (add-edge g a b)))

(defn travel-1 [g v been-there]
  (cond
    ;; if g doesn't contain v it means node v doesn't
    ;; have any edge going out, so we're just adding
    ;; it to set
    (not (contains? g v)) [v]
    (and
     (contains? g v)
     (not ((into #{} been-there) v)))
    (let [been-there (conj been-there v)] (reduce
                                           #(clojure.set/union %1 %2)
                                           been-there
                                           (map #(travel-1 g % been-there) (g v))))
    (and
     (contains? g v)
     ((into #{} been-there) v)) [v]))

(travel-1 g 70 [])
(travel-1 g 70 #{})

(defn nodes [g]
  (into
   #{}
   (concat
    (keys g)
    (reduce concat [] (vals g)) )))

(nodes g)

(defn edges [g]
  (apply concat
         (map-indexed
          (fn [i [a b]] (map (fn [x] [a x]) b))
          g)))

(edges g)

(defn successors [g v]
  (g v))

(successors g 100)

(defn predecessors [g v]
  (map first (filter (fn [[a b]] (some #{v} b)) g)))

(predecessors g 8)

(defn -parent [parent g]
  (ffirst
   (filter (fn [[_ b]] (some #{parent} b)) g)))

(defn path [v1 v2 g]
  (loop [parent (-parent v2 g)
         p (conj nil v2)]
    (if (= v1 parent)
      (conj p parent)
      (recur (-parent parent g) (conj p parent)))))

(path 70 130 g)

(def g (->
        (edge 70 100)
        (edge 100 120)
        (edge 100 130)
        (edge 100 140)
        (edge 120 8)
        (edge 140 8)
        (edge 130 8)
        (edge 8 70)))

(defn graph [& rest]
  (reduce (fn [g [a b]] (edge g a b)) {} rest))

(graph [70 100] [100 120] [100 130] [130 8] [8 70])
